File: 01-intro/03-basic-combinatorial-algorithms/03-combinations/recursive/combination-with-repetition.ex
defmodule Combination do
  def compute(elements, slots) do
    base = {elements, slots, Enum.count(elements)}
    extra = {1, 0, []}
    _compute(base, extra)
  end

  def _compute({_, slots, _}, {slot_index, _, _}) when slot_index > slots, do: nil

  def _compute({elements, slots, elements_count}, {slot_index, after_index, output}) do
    for next_slot_index <- after_index..(elements_count - 1) do
      append_to_output = Enum.at(elements, next_slot_index)
      not_nil = not is_nil(append_to_output)

      if not_nil do
        output = output ++ [append_to_output]

        if slot_index == slots do
          output |> IO.inspect()
        else
          base = {elements, slots, elements_count}
          extra = {slot_index + 1, next_slot_index, output}
          _compute(base, extra)
        end
      end
    end
  end
end

Combination.compute(["a", "b", "c", "d"], 2)
# Combination.compute(["a", "b", "c", "d", "e"], 3)
# IO.inspect(Combination.compute(["a", "b", "c", "d"], 2))
# IO.inspect(Combination.compute(["a", "b", "c", "d"], 2))
# IO.inspect(Combination.compute([0, 1, 2, 3, 4], 2))
# IO.inspect(Combination.compute([0, 1, 2], 3))
# IO.inspect(Combination.compute([0, 1, 2, 3, 4], 3))
# IO.inspect(Combination.compute(["a", "b", "c"], 3))
# Combination.compute(["a", "b", "c"], 2)

File: 01-intro/03-basic-combinatorial-algorithms/03-combinations/recursive/combination-with-repetition.py
class Combination(object):
    elements = []
    elements_count = 0
    slots = 0
    buffer = {}

    def __init__(self, elements: list, slots: int):
        self.elements = elements
        self.elements_count = len(elements)
        self.slots = slots

    def compute(self, slot_index = 1, after = 0):
        if slot_index > self.slots:
            return

        for j in range(after, self.elements_count):
            self.buffer[slot_index - 1] = self.elements[j]
            if slot_index == self.slots:
                print(list(self.buffer.values()))
            self.compute(slot_index + 1, j)


Combination(["a", "b", "c", "d"], 2).compute()
# Combination(["a", "b", "c", "d", "e"], 3).compute()